src/cfml/system/endpoints/S3.cfc

```diff
@@ -0,0 +1,193 @@
+        ].toList(chr(10));
+
+        // string to sign
+        var stringtoSign = [
+            'AWS4-HMAC-SHA256',
+            isoTime,
+            isoTime.left( 8 ) & '/' & bucketRegion & '/s3/aws4_request',
+            hash(canonicalRequest, 'SHA-256').lcase()
+        ].toList(chr(10));
+
+        // signature
+        var signingKey = binaryDecode(hmac(isoTime.left(8), 'AWS4' & awsSettings.awsSecretKey, 'hmacSHA256', 'utf-8'), 'hex');
+        signingKey = binaryDecode(hmac(bucketRegion, signingKey, 'hmacSHA256', 'utf-8'), 'hex');
+        signingKey = binaryDecode(hmac('s3', signingKey, 'hmacSHA256', 'utf-8'), 'hex');
+        signingKey = binaryDecode(hmac('aws4_request', signingKey, 'hmacSHA256', 'utf-8'), 'hex');
+        var signature = hmac(stringToSign, signingKey, 'hmacSHA256', 'utf-8').lcase();
+
+        qs &= '&X-Amz-Signature=' & signature;
+        return '//' & host & path & '?' & qs;
+    }
+
+    private function resolveBucketRegion(bucket, defaultRegion) {
+        var req = '';
+        cfhttp(url='https://s3.#defaultRegion#.amazonaws.com', method='HEAD', result='req', redirect=false) {
+            cfhttpparam(type='header', name='Host', value='#bucket#.s3.amazonaws.com');
+        }
+        return req.responseheader['x-amz-bucket-region'];
+    }
+
+    private function resolveAwsSettings() {
+        var settings = {
+            awsKey: systemSettings.getSystemSetting('AWS_ACCESS_KEY_ID', ''),
+            awsSecretKey: systemSettings.getSystemSetting('AWS_SECRET_ACCESS_KEY', ''),
+            sessionToken: systemSettings.getSystemSetting('AWS_SESSION_TOKEN', ''),
+            defaultRegion: systemSettings.getSystemSetting('AWS_DEFAULT_REGION', ''),
+            profile: systemSettings.getSystemSetting('AWS_PROFILE', 'default'),
+            configFile: systemSettings.getSystemSetting('AWS_CONFIG_FILE', '~/.aws/config'),
+            credentialsFile: systemSettings.getSystemSetting('AWS_SHARED_CREDENTIALS_FILE', '~/.aws/credentials')
+        };
+
+        if (!settings.awsKey.len() || !settings.awsSecretKey.len()) {
+            var credentials = resolveCredentials(settings.credentialsFile, settings.profile);
+            settings.append(credentials);
+        }
+
+        if (!settings.defaultRegion.len()) {
+            var config = loadSettingsFile(settings.configFile);
+            settings.defaultRegion = config[settings.profile].region ?: 'us-east-1';
+        }
+
+        return settings;
+    }
+
+    private function resolveCredentials(credentialsFile, profile) {
+        // check for an aws credentials file for current user
+        var credentials = loadSettingsFile(credentialsFile);
+        try {
+            return {
+                awsKey: credentials[profile].aws_access_key_id,
+                awsSecretKey: credentials[profile].aws_secret_access_key,
+                sessionToken: credentials[profile].aws_session_token ?: ''
+            };
+        } catch(any e) {
+            // pass
```

Comment:
You should pass through the `verbose` flag and log something in these catch blocks so a user can debug how and where their credentials are getting looked up when they use `--verbose` in their install.